src/nvme/types.h

```diff
@@ -7643,4 +7643,95 @@ enum nvme_io_mgmt_send_mo {
+ *	      in the RNUMZRWA field of this data structure. If cleared to ‘0’, then
+ *	      no ZRWA resources are allocated to the namespace to be created. If
+ *	      the ZRWASUP bit is cleared to ‘0’, then this field shall be ignored
+ *	      by the controller.
+ * @rar:       Requested Active Resources specifies the number of active
+ *	      resources to be allocated to the created namespace.
+ * @ror:       Requested Open Resources specifies the number of open resources
+ *	      to be allocated to the created namespace.
+ * @rnumzrwa:  Requested Number of ZRWA Resources specifies the number of ZRWA
+ *	      resources to be allocated to the created namespace.
+ */
+struct nvme_ns_mgmt_host_sw_specified_zns {
+	__u8			znsco;
+	__u32			rar;
+	__u32			ror;
+	__u32			rnumzrwa;
+};
+
+
+/**
+ * struct nvme_ns_mgmt_host_sw_specified - Namespace management Host Software
+ * Specified Fields.
+ * @nsze:     Namespace Size indicates the total size of the namespace in
+ *	      logical blocks. The number of logical blocks is based on the
+ *	      formatted LBA size.
+ * @ncap:     Namespace Capacity indicates the maximum number of logical blocks
+ *	      that may be allocated in the namespace at any point in time. The
+ *	      number of logical blocks is based on the formatted LBA size.
+ * @rsvd16:   Reserved
+ * @flbas:    Formatted LBA Size, see &enum nvme_id_ns_flbas.
+ * @rsvd27:   Reserved
+ * @dps:      End-to-end Data Protection Type Settings, see
+ *	      &enum nvme_id_ns_dps.
+ * @nmic:     Namespace Multi-path I/O and Namespace Sharing Capabilities, see
+ *	      &enum nvme_id_ns_nmic.
+ * @rsvd31:   Reserved
+ * @anagrpid: ANA Group Identifier indicates the ANA Group Identifier of the
+ *	      ANA group of which the namespace is a member.
+ * @rsvd96:   Reserved
+ * @nvmsetid: NVM Set Identifier indicates the NVM Set with which this
+ *	      namespace is associated.
+ * @endgid:   Endurance Group Identifier indicates the Endurance Group with
+ *	      which this namespace is associated.
+ * @rsvd104:  Reserved
+ * @lbstm:    Logical Block Storage Tag Mask Identifies the mask for the
+ *        Storage Tag field for the protection information
+ * @nphndls:  Number of Placement Handles specifies the number of Placement
+ *        Handles included in the Placement Handle List
+ * @rsvd394:  Reserved
+ * @rsvd499:  Reserved for I/O Command Sets that extend this specification.
+ * @zns:      rsvd499( Zoned Namespace Command Set specific field )
+ *        see &struct nvme_ns_mgmt_host_sw_specified_zns.
+ * @phndl:    Placement Handle Associated RUH : This field specifies the Reclaim
+ *        Unit Handle Identifier to be associated with the Placement Handle
+ *        value. If the Flexible Data Placement capability is not supported or
+ *        not enabled in specified Endurance Group, then the controller shall
+ *        ignore this field.
+ * @rsvd768:   Reserved
+ */
+struct nvme_ns_mgmt_host_sw_specified {
```

Comment:
As far I understood TP4146 extends the existing `struct nvme_id_ns` data structure. That means you should actually update the `struct nvme_id_ns` accordingly, no? That would also solve the API breakage as you don't have to change the type.

Reply:
Spec mentioned that As "Fields that are not a subset of the Identify Namespace data structure."
That's why I updated as this, do you think it's better to be included in id_ns?

Reply:
Oh, I missed this part. Need to read up on this. I really hope they didn't come up with polymorphism...

Reply:
In the base spec `Figure 300: Namespace Management`,it says that the first 512 bytes are CSI specific (see I/O command set spec). I understand that this is for CSI zns and previous type was for CSI NVM. Anyway, it looks like the type `struct nvme_id_ns` is not completely correct either. It should be something for `Figure 105: Namespace Management – Host Software Specified Fields`. I suspect that `struct nvme_ns_id` was used because it is very close to the type we should use here.

Right, so I suggest that you just extend the `struct nvme_ns_mgmt_args` in a backwards compatible way. I don't think we should fix the existing type at this point for CSI NVM. This boat has sailed.

@birkelund just to keep you in the loop here. You added the other bits for FDP.

Reply:
I understand what you mean. (remain API format for backwards compatibility)
Then in this case, What's better ?
1. define struct nvme_ns_mgmt_host_sw_specified and send as (struct nvme_ns_id *)
2. define I/O Command Set specific structure only, and copy to 384 addr on (struct nvme_ns_id *)

Reply:
I was thinking on doing something like this here: 9350e73e302a ("libnvme: 64-bit Reference Tags and TP-4068 changes")